File: ios/Runner/Tajweed/SentencePieceTokenizer.swift
import Foundation

/// Lightweight SentencePiece-compatible encode/decode using `tokens.txt`.
/// Decode matches SP piece join; encode uses greedy longest-match over the vocab
/// (adequate for Phase 2; golden tests should compare diacritic-insensitive text).
final class SentencePieceTokenizer {
  private var idToPiece: [String] = []
  private var pieceToId: [String: Int] = [:]
  private var piecesByLength: [String] = []

  init(tokensFileURL: URL) throws {
    let text = try String(contentsOf: tokensFileURL, encoding: .utf8)
    var pieces: [String] = []
    for line in text.split(whereSeparator: \.isNewline) {
      let parts = line.split(separator: " ", maxSplits: 1, omittingEmptySubsequences: false)
      guard parts.count >= 1 else { continue }
      let piece = String(parts[0])
      let id = pieces.count
      pieces.append(piece)
      pieceToId[piece] = id
    }
    guard !pieces.isEmpty else {
      throw TajweedNativeError(TajweedErrorCode.modelLoadFailed, "Empty tokens.txt")
    }
    idToPiece = pieces
    piecesByLength = pieces.sorted { $0.count > $1.count }
  }

  func decode(_ ids: [Int]) -> String {
    var raw = ""
    for id in ids {
      guard id >= 0, id < idToPiece.count else { continue }
      raw += idToPiece[id]
    }
    // SentencePiece: ▁ marks word boundary / space
    return raw.replacingOccurrences(of: "▁", with: " ").trimmingCharacters(in: .whitespaces)
  }

  func encode(_ text: String) -> [Int] {
    // Mirror SP: leading ▁ for word starts.
    var normalized = text.trimmingCharacters(in: .whitespacesAndNewlines)
    if !normalized.isEmpty, !normalized.hasPrefix("▁") {
      normalized = "▁" + normalized.replacingOccurrences(of: " ", with: "▁")
    }
    var ids: [Int] = []
    var i = normalized.startIndex
    while i < normalized.endIndex {
      var matched = false
      for piece in piecesByLength {
        if normalized[i...].hasPrefix(piece), let id = pieceToId[piece] {
          ids.append(id)
          i = normalized.index(i, offsetBy: piece.count)
          matched = true
          break
        }
      }
      if !matched {
        // Skip unknown scalar
        i = normalized.index(after: i)
      }
    }
    return ids
  }

  func piece(for id: Int) -> String {
    guard id >= 0, id < idToPiece.count else { return "" }
    return idToPiece[id].replacingOccurrences(of: "▁", with: "")
  }

  /// True when `id`'s raw piece carries the SentencePiece "▁" word-start marker — i.e. this
  /// piece begins a new Quran word rather than continuing the previous one. Used to group
  /// per-piece forced-align intervals back into whole-word tokens.
  func startsNewWord(_ id: Int) -> Bool {
    guard id >= 0, id < idToPiece.count else { return false }
    return idToPiece[id].hasPrefix("▁")
  }
}

File: android/app/src/main/kotlin/com/app/deenly/deenly/tajweed/SentencePieceTokenizer.kt
package com.rnr.deenfocus.tajweed

import java.io.File

/**
 * Lightweight SentencePiece-compatible encode/decode using `tokens.txt`.
 * Port of ios/Runner/Tajweed/SentencePieceTokenizer.swift — greedy longest-match
 * over the vocab (adequate for Phase 3; golden tests should compare
 * diacritic-insensitive text).
 */
class SentencePieceTokenizer(tokensFile: File) {
    private val idToPiece: List<String>
    private val pieceToId = HashMap<String, Int>()
    private val piecesByLength: List<String>

    init {
        val pieces = ArrayList<String>()
        tokensFile.readLines(Charsets.UTF_8).forEach { line ->
            if (line.isEmpty()) return@forEach
            val spaceIdx = line.indexOf(' ')
            val piece = if (spaceIdx >= 0) line.substring(0, spaceIdx) else line
            val id = pieces.size
            pieces.add(piece)
            pieceToId[piece] = id
        }
        if (pieces.isEmpty()) {
            throw TajweedNativeException(TajweedErrorCode.MODEL_LOAD_FAILED, "Empty tokens.txt")
        }
        idToPiece = pieces
        piecesByLength = pieces.sortedByDescending { it.length }
    }

    fun decode(ids: List<Int>): String {
        val sb = StringBuilder()
        for (id in ids) {
            if (id in idToPiece.indices) sb.append(idToPiece[id])
        }
        return sb.toString().replace("\u2581", " ").trim()
    }

    fun encode(text: String): List<Int> {
        var normalized = text.trim()
        if (normalized.isNotEmpty() && !normalized.startsWith("\u2581")) {
            normalized = "\u2581" + normalized.replace(" ", "\u2581")
        }
        val ids = ArrayList<Int>()
        var i = 0
        while (i < normalized.length) {
            var matched = false
            for (piece in piecesByLength) {
                if (normalized.startsWith(piece, i)) {
                    ids.add(pieceToId.getValue(piece))
                    i += piece.length
                    matched = true
                    break
                }
            }
            if (!matched) i += 1
        }
        return ids
    }

    fun piece(id: Int): String {
        if (id !in idToPiece.indices) return ""
        return idToPiece[id].replace("\u2581", "")
    }

    /** True when [id]'s raw piece carries the SentencePiece "▁" word-start marker —
     * i.e. this piece begins a new Quran word rather than continuing the previous one.
     * Used to group per-piece forced-align intervals back into whole-word tokens. */
    fun startsNewWord(id: Int): Boolean {
        if (id !in idToPiece.indices) return false
        return idToPiece[id].startsWith("\u2581")
    }
}
